src/librustc_resolve/lib.rs

```diff
@@ -1270,6 +1269,24 @@ impl<'a> NameBinding<'a> {
     fn descr(&self) -> &'static str {
         if self.is_extern_crate() { "extern crate" } else { self.def().kind_name() }
     }
+
+    // Suppose that we resolved macro invocation with `invoc_id` to binding `binding` at some
+    // expansion round `max(invoc_id, binding)` when they both emerged from macros.
+    // Then this function returns `true` if `self` may emerge from a macro *after* that
+    // in some later round and screw up our previously found resolution.
+    fn may_appear_after(&self, invoc_id: Mark, binding: &NameBinding) -> bool {
+        // self > max(invoc_id, binding) => !(self <= invoc_id || self <= binding)
+        // Expansions are partially ordered, so "may appear after" is an inversion of
+        // "certainly appears before or simultaneously" and includes unordered cases.
+        let self_parent_expansion = self.expansion;
+        let other_parent_expansion = binding.expansion;
+        let invoc_parent_expansion = invoc_id.parent();
+        let certainly_before_other_or_simultaneously =
+            other_parent_expansion.is_descendant_of(self_parent_expansion);
+        let certainly_before_invoc_or_simultaneously =
+            invoc_parent_expansion.is_descendant_of(self_parent_expansion);
```

Comment:
I'm not sure I understand this. I think this is saying that the `self` would legally be shadowed by any resolution that happens in "a child" macro.. but is that right?  I'm probably missing some bit of context here. I guess I'm imagining that this would apply in a case like:

```rust
macro m { .. }
macro foo {
    () => m!()
}
foo!() { }
```

Here, the invocation `foo!` is a "sub-expansion" of the one that defined `m` -- but this seems to suggest that there hence `m` the decision to use `m` cannot be shadowed. I guess that means that (e.g.) we have "definition 1" here?

```rust
macro m { .. } // definition 1

macro gen_m { () => macro m { ... } }
gen_m!(); // generates definition 2, possibly later

macro foo {
    () => m!() // resolves to definition 1
}
foo!() { }
```